docs/architecture.md

```diff
@@ -0,0 +1,473 @@
+    type GossipedBlockHeader: NetworkData<SealedFuelBlockHeader>;
+    
+    async fn fetch_best_network_block_header() -> Result<Self::SealedHeaderResponse>;
+    async fn fetch_blocks(query: Range<BlockHeight>) -> Result<Self::BlockResponse>;
+    // punish the sender for providing an invalid block header
+    fn report_invalid_block_header(invalid_header: &Self::SealedHeaderResponse) -> Result<()>;
+    // punish the sender for providing a set of blocks that aren't valid
+    fn report_invalid_blocks(invalid_blocks: &Self::BlockResponse) -> Result<()>;
+    // await a newly produced block from the network (similar to stream.next())
+    async fn next_gossiped_block_header() -> Result<Self::SealedFuelBlockHeader>;
+    // notify the p2p network whether to continue gossiping this message to others or
+    // punish the peer that sent it
+    fn notify_gossip_block_validity(message: &Self::GossipedBlockHeader, validity: GossipValidity);
+}
+
+// Generic wrapper for data received from the network
+pub trait NetworkData<D> {
+    // Transfer ownership of the provided data without consuming other message metadata
+    fn take_data(&mut self) -> Option<D>;
+}
+
+pub enum GossipValidity {
+    // Report whether the gossiped message is valid and safe to rebroadcast
+    Accept,
+    // Ignore the received message and prevent further gossiping
+    Ignore,
+    // Punish the gossip sender for providing invalid 
+    // (or malicious) data and prevent further gossiping
+    Invalid,
+}
+
+trait BlockImporter {
+    // commit a sealed block to the chain
+    async fn commit(block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Consensus {
+    // check with the consensus layer whether a block header passes consensus rules
+    async fn validate_sealed_block_header(block: SealedFuelBlockHeader) -> Result<()>;
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    type GossipedTransaction: NetworkData<Transaction>;
+    // Gossip broadcast a transaction inserted via API.
+    async fn broadcast_transaction(transaction: Transaction);
+    // Await the next transaction from network gossip (similar to stream.next()).
+    async fn next_gossiped_transaction(&self) -> Self::GossipedTransaction;
```

Comment:
Won't this need an error case?

Reply:
Or maybe that's handled in the `NetworkData` trait?

Reply:
Possibly? I was thinking this could be consumed in a similar manner to [steam.next()](https://docs.rs/futures/latest/futures/stream/trait.Stream.html#tymethod.poll_next). Since this should never close while the service is running, I didn't think it'd need an option either.

Reply:
Yep so it's basically wait forever until the next transaction arrives. Maybe we could log errors internally or something.